plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/rule/InMemoryRuleProcessingService.java

```diff
@@ -0,0 +1,125 @@
+     *                           matching rule
+     * @return a label if there is unique label otherwise empty
+     */
+    public Optional<String> evaluateLabel(List<AttributeExtractor<String>> attributeExtractors) {
+        Optional<String> result = Optional.empty();
+        for (AttributeExtractor<String> attributeExtractor : attributeExtractors) {
+            AttributeValueStore valueStore = attributeExtractor.getAttribute().getValueStore();
+            for (String value : attributeExtractor.extract()) {
+                Optional<String> possibleMatch = valueStore.get(value);
+
+                if (possibleMatch.isEmpty()) {
+                    return Optional.empty();
+                }
+
+                if (result.isEmpty()) {
+                    result = possibleMatch;
+                } else {
+                    boolean isThePossibleMatchEqualResult = possibleMatch.get().equals(result.get());
+                    if (!isThePossibleMatchEqualResult) {
+                        return Optional.empty();
+                    }
+                }
+            }
+        }
+        return result;
+    }
+
+    private static abstract class RuleProcessingOperation {
+        protected final Rule rule;
+
+        public RuleProcessingOperation(Rule rule) {
+            this.rule = rule;
+        }
+
+        void perform() {
+            final Rule.Feature feature = rule.getFeature();
+            final String label = rule.getLabel();
+
+            for (Map.Entry<Rule.Attribute, List<String>> attributeEntry : rule.getAttributeMap().entrySet()) {
+                processAttributeEntry(attributeEntry);
+            }
+        }
+
+        protected abstract void processAttributeEntry(Map.Entry<Rule.Attribute, List<String>> attributeEntry);
+    }
+
+    private static class DeleteRuleOperation extends RuleProcessingOperation {
+        public DeleteRuleOperation(Rule rule) {
+            super(rule);
+        }
+
+        @Override
+        protected void processAttributeEntry(Map.Entry<Rule.Attribute, List<String>> attributeEntry) {
+            Rule.Attribute attribute = attributeEntry.getKey();
+            assert attribute.getValueStore() != null;
+
+            for (String value : attributeEntry.getValue()) {
+                attribute.getValueStore().remove(value);
+            }
+        }
+    }
+
+    private static class AddRuleOperation extends RuleProcessingOperation {
+        public AddRuleOperation(Rule rule) {
+            super(rule);
+        }
+
+        @Override
+        protected void processAttributeEntry(Map.Entry<Rule.Attribute, List<String>> attributeEntry) {
+            Rule.Attribute attribute = attributeEntry.getKey();
+            assert attribute.getValueStore() != null;
+
+            for (String value : attributeEntry.getValue()) {
+                attribute.getValueStore().add(value, this.rule.getLabel());
```

Comment:
Why not have reference to the rule itself, instead of `String` label?

Reply:
I believe there is no added advantage of adding rule reference here at this place. Do you think that will add any additional value ?